File: 51_N_Queens.py
# Leetcode problem link : https://leetcode.com/problems/n-queens/
# Time Complexity:    O(n!)
# Space Complexity:   O(n*n) for board
#  Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Your code here along with comments explaining your approach
'''
    1. Place one queen in the first row.
    2. Iterate through all the columns and check if queen can be placed. If yes proceed to next row. Else, backtrack and place queen in next column for previous row and repeat.
    3. If we have reached last row+1 then all the queens were placed succesfully, add it to output.

'''

class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        if not n:
            return []
        board = [['.' for _ in range(n)] for _ in range(n)]
        output = []
        self.backtrack(n,output,0,board)
        return output
    
    def backtrack(self,n,output,i,board):
        if i == n:
            self.addToOutput(board,output)
            return
        for j in range(0,len(board)):
            if(self.canPlaceQueen(board,i,j)):
                board[i][j] = 'Q'
                self.backtrack(n,output,i+1,board)
                board[i][j] = '.'
    
    def canPlaceQueen(self,board,i,j):
        r = i
        c = j
        
        while r >= 0:
            if board[r][c] == 'Q':
                return False
            r -= 1
        
        r = i
        
        while r >= 0 and c >= 0:
            if board[r][c] == 'Q':
                return False
            r -= 1
            c -= 1
        
        r = i
        c = j
        
        while r >= 0 and c < len(board):
            if board[r][c] == 'Q':
                return False
            r -= 1
            c += 1
        
        return True
    
    def addToOutput(self,board,output):
        temp = []
        for i in range(len(board)):
            row = ""
            for j in range(len(board)):
                row += board[i][j]
            temp.append(row)
        output.append(temp)
            
        

File: _51_N-Queens.java
// Time Complexity : o(n!))
// Space Complexity : o(n*n)) length of recursive stack
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach
// For every row place a queen and call recusion for next queen from next row starting if not possible backtrack

class Solution {

    Set<Integer> row;
    Set<Integer> column;
    Set<Integer> diagonalright;
    Set<Integer> diagonalleft;

    List<List<String>> res;

    public List<List<String>> solveNQueens(int n) {
        row = new HashSet<>();
        column = new HashSet<>();
        diagonalright = new HashSet<>();
        diagonalleft = new HashSet<>();

        res = new ArrayList<>();

       char[][] board = new char[n][n];
        for(char[] row : board){
            Arrays.fill(row, '.');
        }
        dfs(n, 0, 0, 0, board);
        return res;

    }

    public void dfs(int n, int count, int x, int y, char[][] board) {

        if(n == count){
            copyBoard(board);
            return;
        }
        int i = count;
        //if we use for loop fist queen will be tried to b placed in all cells one by one but we dont need to place 1st queen except row 1
        //same for other queens too
        //for(int i=x; i<n; i++){
              for(int j=y; j<n; j++){

                  if(isValidPlacement(i,j)){

                      row.add(i);
                      column.add(j);
                      diagonalright.add(i-j);
                      diagonalleft.add(i+j);

                      board[i][j] = 'Q';

                      dfs(n, count+1, i+1, 0, board);

                      board[i][j] = '.';

                      row.remove(i);
                      column.remove(j);
                      diagonalright.remove(i-j);
                      diagonalleft.remove(i+j);

                  }

              }
        //}

    }

    public boolean isValidPlacement(int i, int j){
        if( row.contains(i)   ||
            column.contains(j) ||
            diagonalright.contains(i-j)  ||
            diagonalleft.contains(i+j)  ){
                return false;
        }
        return true;
    }


    public void copyBoard(char[][] board){
        List<String> boardList = new ArrayList<>();
        for(char[] row : board){
            boardList.add(new String(row));
        }
        res.add(boardList);
    }


    public int getCellId(int n, int i, int j){
        return (i%n)*n + j%n;
    }


}
